### PR TITLE
Skip lambda arguments for Throws variants in CA1806

### PR DESCRIPTION
**Pull request by 333fred**
Fixes https://github.com/dotnet/roslyn-analyzers/issues/1369.

### CODE REVIEW

File: src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs [hunk 1]
```diff
@@ -110,68 +110,193 @@ public override void Initialize(AnalysisContext analysisContext)
...
+            {
+                var methodNames = ImmutableHashSet.Create(new[]
+                {
+                    "Throws",
+                    "Catch",
+                    "DoesNotThrow",
+                    "ThrowsAsync",
+                    "CatchAsync",
+                    "DoesNotThrowAsync"
+                });
```
Comment (ivanbasov) [inline src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs]: Should not we move this HashSet to private static readonly? And the same for the one below.

File: src/Microsoft.CodeQuality.Analyzers/UnitTests/Maintainability/DoNotIgnoreMethodResultsTests.cs [hunk 2]
```diff
@@ -288,6 +394,126 @@ End Module
...
+        Console.WriteLine(Me)
+        Dim sample As String = ""Sample""
+        sample.ToLower(CultureInfo.InvariantCulture)
+        Return
+    End Sub
+End Class", MSTestExpectedExceptionAttribute.VisualBasic },
+                GetBasicStringCreationResultAt(11, 9, "ThrowsException", "ToLower"));
+        }
+
+
```
Comment (ivanbasov) [inline src/Microsoft.CodeQuality.Analyzers/UnitTests/Maintainability/DoNotIgnoreMethodResultsTests.cs]: nit: remove empty line

File: src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs [hunk 3]
```diff
@@ -110,68 +110,193 @@ public override void Initialize(AnalysisContext analysisContext)
...
 
-            analysisContext.RegisterOperationBlockStartAction(osContext =>
+            analysisContext.RegisterCompilationStartAction(compilationContext =>
             {
-                var method = osContext.OwningSymbol as IMethodSymbol;
-                if (method == null)
-                {
-                    return;
-                }
+                INamedTypeSymbol expectedExceptionType = WellKnownTypes.ExpectedException(compilationContext.Compilation);
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs]: >expectedExceptionType [](start = 33, length = 21)

Consider even caching well known types for `NUnit.Framework.Assert` and `Xunit.Assert`, so you can bail out upfront from `ShouldSkipAnalyzing` if none of these types are defined.

File: src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs (hunk 1 above)
Comment (mavasani) [reply src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs]: +1

---
In reply to: [162737270](https://github.com/dotnet/roslyn-analyzers/pull/1539#discussion_r162737270) [](ancestors = 162737270)

File: src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs [hunk 4]
```diff
@@ -110,68 +110,193 @@ public override void Initialize(AnalysisContext analysisContext)
...
+                {
+                    if (!statement.IsImplicit)
+                    {
+                        return false;
+                    }
+                }
+                else
+                {
+                    continue;
+                }
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs]: Unnecessary?

File: src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs [hunk 5]
```diff
@@ -110,68 +110,193 @@ public override void Initialize(AnalysisContext analysisContext)
...
                     }
 
-                    if (rule != null)
+                    osContext.RegisterOperationAction(opContext =>
                     {
-                        Diagnostic diagnostic = Diagnostic.Create(rule, expression.Syntax.GetLocation(), method.Name, targetMethodName);
-                        opContext.ReportDiagnostic(diagnostic);
-                    }
-                }, OperationKind.ExpressionStatement);
+                        if (ShouldSkipAnalyzing(opContext, expectedExceptionType))
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs]: >ShouldSkipAnalyzing [](start = 28, length = 19)

Should we invoke this method after we have determined the method result is unused (just before reporting the diagnostic)? That might be more efficient..

File: src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs [hunk 6]
```diff
@@ -110,68 +110,193 @@ public override void Initialize(AnalysisContext analysisContext)
...
+                {
+                    parentArgument = parentArgument.Parent;
+                } while (parentArgument != null && parentArgument.Kind != OperationKind.Argument);
+
+                if (parentArgument == null)
+                {
+                    return false;
+                }
+
+                IArgumentOperation argumentOperation = (IArgumentOperation)parentArgument;
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs]: Looks like a good candidate for an extension method `TOperation GetAncestor<TOperation>(this IOperation operation, OperationKind ancestorKind) where TOperation: IOperation` for AnalyzerUtilities?

**Comment by 333fred** [comment]
@mavasani @ivanbasov addressed feedback.

File: src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs [hunk 7]
```diff
@@ -110,66 +130,163 @@ public override void Initialize(AnalysisContext analysisContext)
...
+                }
+
+                IMethodSymbol methodSymbol = (IMethodSymbol)operationContext.ContainingSymbol;
+
+                return methodSymbol.GetAttributes().Any(attr => attr.AttributeClass == expectedExceptionType);
+            }
+            else
+            {
+                IArgumentOperation argumentOperation = enclosingBlock.GetAncestor<IArgumentOperation>(OperationKind.Argument);
+                return IsNUnitThrowsArgument(argumentOperation.Parameter) || IsXunitThrowsArgument(argumentOperation.Parameter);
```
Comment (mavasani) [inline src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs]: Do we need a null check for `argumentOperation`?

File: src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs [hunk 8]
```diff
@@ -43,6 +44,25 @@ public sealed class DoNotIgnoreMethodResultsAnalyzer : DiagnosticAnalyzer
...
+
+        private static readonly ImmutableHashSet<string> s_xUnitMethodNames = ImmutableHashSet.Create(
+            new[] {
+                "Throws",
+                "ThrowsAsync",
+                "ThrowsAny",
+                "ThrowsAnyAsync",
+            });
+
+
```
Comment (ivanbasov) [inline src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs]: nit: remove line

**Pull request merged**